Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -480,60 +480,35 @@ internal final class AsyncServerHandler<
   @Sendable
   @usableFromInline
   internal func invokeUserHandler(
-    sequence: AsyncSequenceProducer,
-    sequenceSource: AsyncSequenceProducer.Source,
-    responseStreamWriter: AsyncWriter<AsyncResponseStreamWriterDelegate<Response>>,
+    requestSequence: AsyncSequenceProducer.NewSequence,
+    responseWriter: NIOAsyncWriter<
+      (Response, Compression),
+      GRPCAsyncWriterSinkDelegate<(Response, Compression)>
+    >,
     callContext: GRPCAsyncServerCallContext
-  ) async throws {
+  ) async {
     defer {
       // It's possible the user handler completed before the end of the request stream. We
       // explicitly finish it to drop any unconsumed inbound messages.
-      sequenceSource.finish()
+      requestSequence.source.finish()
     }
 
     do {
-      let requestStream = GRPCAsyncRequestStream(sequence)
-      let responseStream = GRPCAsyncResponseStreamWriter(wrapping: responseStreamWriter)
-      try await self.userHandler(requestStream, responseStream, callContext)
+      let grpcRequestStream = GRPCAsyncRequestStream(requestSequence.sequence)
+      let grpcResponseStreamWriter = GRPCAsyncResponseStreamWriter(wrapping: responseWriter)
+      try await self.userHandler(grpcRequestStream, grpcResponseStreamWriter, callContext)
 
-      // Done successfully. Queue up and send back an 'ok' status.
-      try await responseStreamWriter.finish(.ok)
+      responseWriter.finish()
     } catch {
-      // Drop pending writes as we're on the error path.
-      await responseStreamWriter.cancel(withError: error)
-
       if let thrownStatus = error as? GRPCStatus, thrownStatus.isOk {
-        throw GRPCStatus(code: .unknown, message: "Handler threw error with status code 'ok'.")
+        let unknownStatus = GRPCStatus(
+          code: .unknown,
+          message: "Handler threw error with status code 'ok'."
+        )
+        responseWriter.finish(error: unknownStatus)
```

Comment:
Maybe I'm getting my wires crossed, but didn't we discuss elsewhere that we would handle all the error logic wherever we consume the error? I.e. the catch block should just finish the response writer with whatever error was thrown.

Reply:
Hmm yes, and I did move some of the logic in there (`_interceptTermination(_:)` method in this class), but you're right I should probably move this in there as well. Do we want to wrap every error in a `GRPCStatus.unknown` error though? Or should we just rely on the output of `ServerErrorProcessor.processObserverError`, which is all `_interceptTermination(_:)` is currently doing? If we went for the latter, we'd lose the `"Handler threw error with status code 'ok'"` message, don't know how much we care about that.

Reply:
Not every error, no. We want to put all errors through `ServerErrorProcessor.processObserverError` and rely on the output of that. We should make sure we retain the conversion from thrown 'ok' to 'unknown' though, including the message. That should probably be done in `_interceptTermination` or `ServerErrorProcessor.processObserverError`.